test_rmsd.py fails with numpy 2.4 scalar conversion TypeError

<!-- 
If this is your first Issue submitted to the BioPandas Issue Tracker, please review
the code of conduct, which is available at http://rasbt.github.io/biopandas/Code-of-Conduct/. 
-->


<!--
Before submitting a bug, please check the recent Changelog entries at 
https://github.com/rasbt/biopandas/blob/master/docs/sources/CHANGELOG.md
and do a quick search in the Issue Tracker (https://github.com/rasbt/biopandas/issues)
to make sure the issue hasn't been already
addressed.
-->

#### Describe the bug

```test_rmsd.py``` fails with numpy >=2.4

```
==================================== ERRORS ====================================
___________________ ERROR collecting tests/mmtf/test_rmsd.py ___________________
tests/mmtf/test_rmsd.py:29: in <module>
    p1t48.read_mmtf(TESTDATA_1t48)
biopandas/mmtf/pandas_mmtf.py:55: in read_mmtf
    df = self._mmtf_to_df(self.mmtf)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
biopandas/mmtf/pandas_mmtf.py:65: in _mmtf_to_df
    return mmtf_to_df(mmtf_obj)
           ^^^^^^^^^^^^^^^^^^^^
biopandas/mmtf/pandas_mmtf.py:585: in mmtf_to_df
    int(np.argwhere(np.array(model_indices) > ch_idx)[0]) + 1
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   TypeError: only 0-dimensional arrays can be converted to Python scalars
=========================== short test summary info ============================
ERROR tests/mmtf/test_rmsd.py - TypeError: only 0-dimensional arrays can be c...
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
```

<!--
Briefly describe what the bug is.
-->

#### Steps/Code to Reproduce


If I understand correctly, this change from numpy 1.25, which became enforced in numpy 2.4, is the cause:

https://numpy.org/doc/2.0/release/1.25.0-notes.html

> Only ndim-0 arrays are treated as scalars. NumPy used to treat all arrays of size 1 (e.g., np.array([3.14])) as scalars. In the future, this will be limited to arrays of ndim 0 (e.g., np.array(3.14)). The following expressions will report a deprecation warning:

Example:

```
[chemonke@nixos:~/biopandas]$ python3
Python 3.14.3 (main, Feb  3 2026, 15:32:20) [GCC 15.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.__version__
'2.4.4'
>>> import numpy as np
... x = np.arange(4)
... x
... 
array([0, 1, 2, 3])
>>> int(np.argwhere(x>1)[0])
Traceback (most recent call last):
  File "<python-input-3>", line 1, in <module>
    int(np.argwhere(x>1)[0])
    ~~~^^^^^^^^^^^^^^^^^^^^^
TypeError: only 0-dimensional arrays can be converted to Python scalars
>>> int(np.argwhere(x>1)[0][0])
... 
2
>>> 

```
This line still uses the deprecated behavior
`int(np.argwhere(np.array(model_indices) > ch_idx)[0]) + 1`
https://github.com/BioPandas/biopandas/blob/main/biopandas/mmtf/pandas_mmtf.py#L585

which can be fixed by explicitly selecting the scalar index value
`int(np.argwhere(np.array(model_indices) > ch_idx)[0][0]) + 1`


<!--


<!--
Please add a minimal example that can help understand and reproduce the bug.
Ideally, this should be a self-contained code example that can be run
on our computer with easily accessible datasets (ideally, datasets contained
in scikit-learn or BioPandas itself).


Example:

```python
import numpy as np
from biopandas.evaluate import bootstrap


rng = np.random.RandomState(123)
x = rng.normal(loc=5., size=100)
original, std_err, ci_bounds = bootstrap(x, num_rounds=1000, func=np.mean, ci=0.95, seed=123)
print('Mean: %.2f, SE: +/- %.2f, CI95: [%.2f, %.2f]' % (original, 
                                                        std_err, 
                                                        ci_bounds[0],
                                                        ci_bounds[1]))
```

If the code is too long, feel free to put it in a public gist and link
it in the issue: https://gist.github.com
-->

#### Expected Results

<!-- Please paste or describe the expected results.-->

#### Actual Results
<!-- Please paste or specifically describe the actual output or error traceback. -->

#### Versions

Verified on an cloud instance running Ubuntu, so this is not an issue with NixOS being weird.

biopandas 0.5.1
Linux-6.12.47-x86_64-with-glibc2.39
Python 3.14.4 (main, May  1 2026, 05:51:55) [GCC 13.3.0]
NumPy 2.4.6

PRETTY_NAME="Ubuntu 24.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04.4 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo

<!--
Please run the following snippet and paste the output below.
import biopandas; print("biopandas", biopandas.__version__)
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import sklearn; print("Scikit-learn", sklearn.__version__)
import numpy; print("NumPy", numpy.__version__)
import scipy; print("SciPy", scipy.__version__)
-->


<!-- Thanks for contributing! -->

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -1,5 +1,7 @@
 build: false
 
+image: Visual Studio 2019
+
 environment:
   MAMBA_ROOT_PREFIX: C:\micromamba
   matrix:
@@ -13,7 +15,11 @@ init:
   - ECHO %PYTHON_VERSION%
 
 install:
-  - ps: Invoke-WebRequest -Uri "https://github.com/mamba-org/micromamba-releases/releases/latest/download/micromamba-win-64.exe" -OutFile "C:\Windows\micromamba.exe"
+  - ps: |
+      $ProgressPreference = 'SilentlyContinue'
+      [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+      Invoke-WebRequest -Uri "https://github.com/mamba-org/micromamba-releases/releases/latest/download/micromamba-win-64.exe" -OutFile "$env:USERPROFILE\micromamba.exe" -UseBasicParsing
+  - set PATH=%USERPROFILE%;%PATH%
   - micromamba --version
   - micromamba create -q -y -n test-environment --channel=conda-forge mmtf-python numpy scipy pandas pytest looseversion importlib_resources python=%PYTHON_VERSION%
 

--- a/.appveyor.yml
+++ b/.appveyor.yml
@@ -1,30 +1,21 @@
 build: false
 
 environment:
+  MAMBA_ROOT_PREFIX: C:\micromamba
   matrix:
     - PYTHON_VERSION: 3.7
-      MINICONDA: C:\Miniconda3
     - PYTHON_VERSION: 3.8
-      MINICONDA: C:\Miniconda3
     - PYTHON_VERSION: 3.9
-      MINICONDA: C:\Miniconda3
     - PYTHON_VERSION: "3.10"
-      MINICONDA: C:\Miniconda3
     - PYTHON_VERSION: 3.11
-      MINICONDA: C:\Miniconda3
-
 
 init:
-  - ECHO %PYTHON_VERSION% %MINICONDA%
-  - ECHO conda --version
+  - ECHO %PYTHON_VERSION%
 
 install:
-  - set PATH=%MINICONDA%;%MINICONDA%\\Scripts;%PATH%
-  - conda config --set always_yes yes --set changeps1 no
-  - conda update -q conda
-  - conda info -a
-  - conda create -q -n test-environment --channel=conda-forge mmtf-python numpy scipy pandas pytest looseversion importlib_resources python=%PYTHON_VERSION%
-  - activate test-environment
+  - ps: Invoke-WebRequest -Uri "https://github.com/mamba-org/micromamba-releases/releases/latest/download/micromamba-win-64.exe" -OutFile "C:\Windows\micromamba.exe"
+  - micromamba --version
+  - micromamba create -q -y -n test-environment --channel=conda-forge mmtf-python numpy scipy pandas pytest looseversion importlib_resources python=%PYTHON_VERSION%
 
 test_script:
-  - pytest -s -v
+  - micromamba run -n test-environment pytest -s -v

--- a/biopandas/mmcif/pandas_mmcif.py
+++ b/biopandas/mmcif/pandas_mmcif.py
@@ -430,7 +430,11 @@ class PandasMmcif:
         indices = []
 
         residue_number_insertion = (
-            tmp[residue_number_col].astype(str) + tmp["pdbx_PDB_ins_code"]
+            tmp[chain_col].astype(str)
+            + "_"
+            + tmp[residue_number_col].astype(str)
+            + "_"
+            + tmp["pdbx_PDB_ins_code"].fillna("")
         )
 
         for num, ind in zip(residue_number_insertion, np.arange(tmp.shape[0])):

--- a/biopandas/mmcif/pandas_mmcif.py
+++ b/biopandas/mmcif/pandas_mmcif.py
@@ -213,7 +213,9 @@ class PandasMmcif:
         full_df = pd.DataFrame.from_dict(
             data["atom_site"], orient="index"
         ).transpose()
+        mask = full_df.isna()
         full_df = full_df.astype(mmcif_col_types, errors="ignore")
+        full_df = full_df.where(~mask, other=np.nan)
         df["ATOM"] = pd.DataFrame(full_df[full_df.group_PDB == "ATOM"])
         df["HETATM"] = pd.DataFrame(full_df[full_df.group_PDB == "HETATM"])
         try:

--- a/biopandas/mmtf/pandas_mmtf.py
+++ b/biopandas/mmtf/pandas_mmtf.py
@@ -582,7 +582,7 @@ def mmtf_to_df(mmtf_obj: MMTFDecoder) -> pd.DataFrame:
             # data["chain_id"].append([mmtf_obj.chain_name_list[ch_idx]])
             data["chain_id"].append([mmtf_obj.chain_name_list[ch_idx]])
             data["model_id"].append(
-                int(np.argwhere(np.array(model_indices) > ch_idx)[0]) + 1
+                int(np.argwhere(np.array(model_indices) > ch_idx)[0][0]) + 1
             )
             data["record_name"].append(record)
             data["insertion"].append(mmtf_obj.ins_code_list[idx])

--- a/biopandas/mmtf/pandas_mmtf.py
+++ b/biopandas/mmtf/pandas_mmtf.py
@@ -236,7 +236,11 @@ class PandasMmtf(object):
         indices = []
 
         residue_number_insertion = (
-            tmp["residue_number"].astype(str) + tmp["insertion"]
+            tmp["chain_id"].astype(str)
+            + "_"
+            + tmp["residue_number"].astype(str)
+            + "_"
+            + tmp["insertion"].fillna("")
         )
 
         for num, ind in zip(residue_number_insertion, np.arange(tmp.shape[0])):

--- a/biopandas/pdb/pandas_pdb.py
+++ b/biopandas/pdb/pandas_pdb.py
@@ -258,7 +258,7 @@ class PandasPdb(object):
 
         for sec in records:
             t[sec]["element_symbol"] = t[sec][["atom_name", "element_symbol"]].apply(
-                lambda x: x[0][1] if len(x[1]) == 3 else x[0][0], axis=1
+                lambda x: x.iloc[0][1] if len(x.iloc[1]) == 3 else x.iloc[0][0], axis=1
             )
         return t
 

--- a/biopandas/pdb/pandas_pdb.py
+++ b/biopandas/pdb/pandas_pdb.py
@@ -606,7 +606,13 @@ class PandasPdb(object):
         cmp = "placeholder"
         indices = []
 
-        residue_number_insertion = tmp["residue_number"].astype(str) + tmp["insertion"]
+        residue_number_insertion = (
+            tmp["chain_id"].astype(str)
+            + "_"
+            + tmp["residue_number"].astype(str)
+            + "_"
+            + tmp["insertion"].fillna("")
+        )
 
         for num, ind in zip(residue_number_insertion, np.arange(tmp.shape[0])):
             if num != cmp:

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -1,5 +1,6 @@
 # Release Notes ![](img/logos/3eiy_120.png)
 
+- Fix: improves robustness of `amino3to1` residue identification by handling `NaN` insertion codes (NumPy 2.4+ compatibility) and including chain IDs and separators to avoid collisions.
 - Supports `mol` files that have empty lines between blocks, (Via [Ruibin Liu](https://github.com/Ruibin-Liu) PR #[140](https://github.com/BioPandas/biopandas/pull/140#))
 
 The CHANGELOG for the current development version is available at

--- a/tests/mmcif/test_read_mmcif.py
+++ b/tests/mmcif/test_read_mmcif.py
@@ -121,6 +121,10 @@ def test__read_pdb_raises():
     assert_raises(ValueError, expect, run_code_2)
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_fetch_pdb():
     """Test fetch_pdb"""
 
@@ -136,6 +140,10 @@ def test_fetch_pdb():
         assert ppdb.mmcif_path == "https://files.rcsb.org/download/3eiy.cif"
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_fetch_af2():
     """Test fetch_af2"""
     # Test latest release
@@ -327,6 +335,10 @@ def test_get_df():
     assert shape == (857, 21), shape
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_mmcif_pdb_conversion():
     """Tests conversion from mmCIF df to PDB df"""
     # Multichain test

--- a/tests/mmcif/test_read_mmcif.py
+++ b/tests/mmcif/test_read_mmcif.py
@@ -198,16 +198,16 @@ def test__construct_df():
             2,
             "ATOM",
             1,
-            None,
+            float("nan"),
             "A",
             "N",
             "SER",
             1,
             "23",
             1.0,
-            "None",
+            float("nan"),
             1,
-            None,
+            float("nan"),
             "N",
         ],
         index=[
@@ -239,7 +239,11 @@ def test__construct_df():
     # However, if I compare these values directly, they are equal.
     # assert exp.equals(dfs["ATOM"].loc[0, :])
     for k, v in exp.items():
-        assert dfs["ATOM"].loc[0, :][k] == v, k
+        actual = dfs["ATOM"].loc[0, :][k]
+        if pd.isna(v):
+            assert pd.isna(actual), k
+        else:
+            assert actual == v, k
 
 
 def test_read_pdb():

--- a/tests/pdb/test_read_pdb.py
+++ b/tests/pdb/test_read_pdb.py
@@ -118,6 +118,10 @@ def test__read_pdb_raises():
     assert_raises(ValueError, expect, run_code_2)
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_fetch_pdb():
     """Test fetch_pdb"""
 
@@ -136,6 +140,10 @@ def test_fetch_pdb():
         assert ppdb.pdb_path == "https://files.rcsb.org/download/3eiy.pdb"
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_fetch_af2():
     """Test fetch_pdb"""
     # Check latest release

--- a/tests/pdb/test_write_pdb.py
+++ b/tests/pdb/test_write_pdb.py
@@ -15,6 +15,7 @@ import os
 import warnings
 
 import pandas as pd
+import pytest
 
 import tests.pdb.data
 from biopandas.pdb import PandasPdb
@@ -148,6 +149,10 @@ def test_introduce_remark():
     assert len(ppdb.df["ATOM"]) == n_atoms
 
 
+@pytest.mark.skipif(
+    sys.platform == "win32",
+    reason="Windows CI cert store triggers ssl.SSLError: [ASN1: NOT_ENOUGH_DATA]",
+)
 def test_b_factor_shift():
     """Test b_factor shifting one white space when saving the fetched pdb."""
     ppdb = PandasPdb()

